multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/util/UrisApplicationAttributeUpdater.java

```diff
@@ -19,20 +20,22 @@ public UrisApplicationAttributeUpdater(Context context, UpdateStrategy updateStr
     }
 
     @Override
-    protected boolean shouldUpdateAttribute(CloudApplication existingApplication, CloudApplication application) {
-        return !SetUtils.isEqualSet(application.getRoutes(), existingApplication.getRoutes());
+    protected boolean shouldUpdateAttribute(CloudApplication existingApplication, CloudApplicationExtended application) {
+        var routes = getControllerClient().getApplicationRoutes(existingApplication.getGuid());
+        return !SetUtils.isEqualSet(application.getRoutes(), routes);
     }
 
     @Override
-    protected void updateAttribute(CloudApplication existingApplication, CloudApplication application) {
-        Set<CloudRouteSummary> routes = applyUpdateStrategy(existingApplication.getRoutes(), application.getRoutes());
+    protected void updateAttribute(CloudApplication existingApplication, CloudApplicationExtended application) {
+        var existingRoutes = getControllerClient().getApplicationRoutes(existingApplication.getGuid());
+        Set<CloudRoute> routes = applyUpdateStrategy(existingRoutes, application.getRoutes());
         getControllerClient().updateApplicationRoutes(application.getName(), routes);
     }
 
-    private Set<CloudRouteSummary> applyUpdateStrategy(Set<CloudRouteSummary> existingRoutes, Set<CloudRouteSummary> routes) {
+    private Set<CloudRoute> applyUpdateStrategy(List<CloudRoute> existingRoutes, Set<CloudRoute> routes) {
         getLogger().debug(Messages.EXISTING_URIS_0, UriUtil.prettyPrintRoutes(existingRoutes));
         getLogger().debug(Messages.APPLYING_UPDATE_STRATEGY_0_WITH_URIS_1, updateStrategy, UriUtil.prettyPrintRoutes(routes));
-        Set<CloudRouteSummary> result = getElementUpdater().updateSet(existingRoutes, routes);
+        Set<CloudRoute> result = getElementUpdater().updateSet(new HashSet<>(existingRoutes), routes);
```

Comment:
Why existingRoutes is changed to List and then back converted to Set?

Reply:
Because the return type of the `getApplicationRoutes` method from `CloudControllerClient` is a list. But we keep the routes from the MTA module (`CloudApplicationExtended`) in a set, so that is why we need the conversion